File: 104.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Pandigital Fibonacci ends".
The problem is: Given that Fk is the first Fibonacci number for which the first
nine digits AND the last nine digits are 1-9 pandigital, find k.
"""

import math


def main():
    """Pandigital Fibonacci ends"""

    # Constants
    TRUNCATE = 1e10
    LOG_PHI = math.log((1 + math.sqrt(5)) / 2, 10)
    LOG_ROOT5 = math.log(5, 10) / 2

    k = 2
    fib1 = 1
    fib2 = 1
    while True:
        k += 1
        sum = fib1 + fib2
        fib1 = fib2 % TRUNCATE
        fib2 = sum % TRUNCATE
        if not is_pandigital(fib2):
            continue
        log_fibk = k * LOG_PHI - LOG_ROOT5
        fibk = int(pow(10, log_fibk - math.floor(log_fibk)) * pow(10, 8))
        if is_pandigital(fibk):
            break

    print(k)


def is_pandigital(num):
    digits = set(i for i in str(num))
    if len(digits) == len(str(num)):
        return True
    else:
        return False

if __name__ == '__main__':
    main()

File: 104.js
#!/usr/bin/env node

// Pandigital Fibonacci ends

var TRUNCATE = 1e10;
var LOG_PHI = Math.log10((1 + Math.sqrt(5)) / 2);
var LOG_ROOT5 = Math.log10(5) / 2;

var k = 2;
var fib1 = 1;
var fib2 = 1;
while (true) {
    k++;
    sum = fib1 + fib2;
    fib1 = fib2 % TRUNCATE;
    fib2 = sum % TRUNCATE;
    if (!is_pandigital(fib2)) {
        continue;
    }
    var log_fibk = k * LOG_PHI - LOG_ROOT5;
    var fibk = Math.floor(Math.pow(10, log_fibk - Math.floor(log_fibk)) *
        Math.pow(10, 8));
    if (is_pandigital(fibk)) {
        break;
    }
}

console.log(k);

function is_pandigital(num) {
    var digits = unique(num.toString().split(''));
    if (digits.length == num.toString().length) {
        return true;
    } else {
        return false;
    }
}

function unique(array) {
    return array.reduce(function(prev, cur) {
        if (prev.indexOf(cur) < 0) {
            prev.push(cur);
        }
        return prev;
    }, []);
}
